dayatsin/gpu_remapping

dayatsin-amd opened the pull request:
Adding code for GPU remapping on nodes with different topology
Changed BO content dump to use one thread per GPU to speed-up dump
Changed log messages to print amdgpu_plugin instead of kfd_plugin
Renamed plugin entry-point functions to use amdgpu_plugin* instead of kfd_plugin*
Renamed Documentation/kfd_plugin.txt Documentation/amdgpu_plugin.txt

rajbhar commented:
abandon PR

dayatsin-amd committed:
criu/plugin: Add unit tests for GPU remapping

Adding unit tests for GPU remapping code when checkpointing and
restoring on different nodes with different topologies.

dayatsin-amd committed:
criu/plugin: Read and write BO contents in parallel

Implement multi-threaded code to read and write contents of each GPU VRAM
BOs in parallel in order to speed up dumping process when using multiple
GPUs.

Moving opening and closing of file descriptors outside of loop so
that they are only performed once per GPU.

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -461,11 +611,17 @@ int amdgpu_plugin_dump_file(int fd, int id)
 		if (devinfo_bucket_ptr[i].user_gpu_id != devinfo_bucket_ptr[i].actual_gpu_id) {
 			pr_err("Checkpoint-Restore on different node not supported yet\n");
```
You're still printing an error message, but looks like you're not treating it as an error any more?

dayatsin-amd replied:
I do not have this line in my code, I think this line has been completely replaced in a previous PR

fxkamd replied:
It's still in this commit, but gets removed in the next commit.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -804,26 +961,42 @@ int amdgpu_plugin_restore_file(int id)
+
+	args.num_of_devices = e->num_of_gpus;
+
+	devinfo_bucket_ptr = xmalloc(args.num_of_devices * sizeof(struct kfd_criu_devinfo_bucket));
```
A general comment, since you do this a lot: Using sizeof(struct ...) can be error prone if the type changes in the future, or you mix up what's a pointer and what's a structure. This is safer:

    devinfo_bucket_ptr = xmalloc(args.num_of_devices * sizeof(*devinfo_bucket_ptr));

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -804,26 +961,42 @@ int amdgpu_plugin_restore_file(int id)
-			pr_info("amdgpu_plugin: passing drm render fd = %d to driver\n", devinfo_bucket_ptr[i].drm_fd);
+			pr_info("amdgpu_plugin: passing drm render fd = %d to driver\n", devinfo_bucket_ptr[j].drm_fd);
 		}
+		j++;
```
Looks like j == i in all cases. I don't see any way they can be different in this loop. So j is redundant.

fxkamd replied:
OK, so j becomes more meaningful in the next commit. Maybe a comment would help to explain what i and j represent respectively. Or some more meaningful variable names.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.h:
```diff
@@ -0,0 +1,131 @@
+#define TOPO_IOLINK_TYPE_PCIE	2  /* HSA_IOLINKTYPE_PCIEXPRESS */
+#define TOPO_IOLINK_TYPE_XGMI	11 /* HSA_IOLINK_TYPE_XGMI */
+
+#define NODE_IS_GPU(node) (node->gpu_id != 0)
```
Macro arguments require care because they are just text substitutions during pre-processing. Generally we wrap them in parentheses to avoid surprises:

    #define NODE_IS_GPU(node) ((node)->gpu_id != 0)

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.h:
```diff
@@ -0,0 +1,131 @@
+};
+
+/************************************ Global Variables ********************************************/
+struct tp_system src_topology;  /* Valid during dump */
```
Global variable declarations in the header file should use the "extern" keyword. Otherwise each .c file that includes the header will get its own instance of the global variable. And I think you may get linker warnings about duplicate symbols.

Then you also need a definition of the variables in one source file without the "extern" keyword.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+
+	list_add_tail(&id_map->listm, &maps->gpu_maps);
+
+	maps->tail_gpu = &id_map->listm;
```
I'm not sure why you need maps->tail_gpu (and tail_cpu). You can get it with maps->gpu_maps.prev.

dayatsin-amd replied:
See comment in next conversation below

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+void maps_remove(struct device_maps *maps, struct device_maps *remove)
+{
+	if (remove->tail_cpu)
+		list_cut_position(&remove->cpu_maps, &maps->cpu_maps, remove->tail_cpu);
```
I'm not sure what you're trying to do here, but it doesn't look right. Looking at the documentation in the kernel's list.h, you're adding some entries to remove->cpu_maps. But I think that's what you want to remove entries from.

dayatsin-amd replied:
**maps_remove is the "reverse" of maps_append function.**

Example, if you have 3 lists:
list A = A1 - A2 - A3
list B = B1 - B2 - B3
list C = C1 - C2 - C3

Calling:
maps_append(A, B);
maps_append(A, C);

Will result in:
list A = A1 - A2 -A3 - B1 - B2 - B3 - C1 - C2 - C3

But maps_append will also store the location where the lists where glued together in tail_cpu and tail_gpu. So after maps_append(A, B) returns, the glue position will be stored in B.

So afterwards, if we call:

maps_remove(A, B);

Result will be:
list A= A1 - A2 - A3

i.e list A will be cut at the previous glue position stored in B and this result in elements B1 - B2 - B3 - C1 - C2 - C3 to be removed from A. 


**maps_remove implemetation:**
The list_cut_position function will split a list back into two lists at the previous glue position stored during maps_append.
Using the 3 lists in the previous example, after:

 list_cut_position(&remove->cpu_maps, &maps->cpu_maps, remove->tail_cpu)

remove->cpu_maps will contain elements B1 - B2 - B3 - C1 - C2 - C3
and 
maps->cpu_maps will contain elements A1 - A2 - A3

maps_remove then frees elements B1 - B2 - B3 - C1 - C2 - C3 by calling maps_free(remove)


For the maps->gpu_maps, we always add/remove one node at a time inside the map_device() function. But if one GPU has multiple iolinks to multiple CPUs, then for maps->cpu_maps, we would be appending multiple elements to maps->cpu_maps at a time, and if later on that combination fails, we have to remove multiple elements from maps->cpu_maps. This is why I am storing this glue position in maps_append().

fxkamd replied:
Looks to me like you store the "glue position" tail->cpu or tail->gpu in maps_add_cpu/gpu_entry, not maps_append.

It stores the last element added to the list. So the position stored in B would be B3, for example. But the glue position in your examples, as I would understand it, is A3.

I think you're using those lists like a stack during the recursion. You add things on the stack as you descend, and then pop them back off the stack as you come back up. Your maps_append function doesn't really "append" because it adds the new list entries at the start, not the end. Maybe that explains my confusion. With that in mind, the glue position makes a lot more sense. So the actual order of elements after maps_append(A, B) is B1 - B2 - B3 - A1 - A2 - A3, and B3 is where you want to split the list when you pop B with maps_remove(A, B).

I think renaming your maps_append/remove functions to maps_push/pop would go a long way to clearing that up. Also, add a comment explaining that these lists are used as stacks in the recursion.

dayatsin-amd replied:
ACK. Yes, the intention was to make it behave like a stack.

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+
+			list_del(&dest_node->listm_mapping);
+			if (maps_append(maps, &new_maps))
+				return -EINVAL;
```
-EINVAL will be boolean "true".

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+{
+	struct tp_iolink *iolink;
+	list_for_each_entry(iolink, &node->iolinks, listm) {
+		if (iolink->node_to->id == node_id && iolink->type == type)
```
Any reason why you're not using iolink->node_to_id here?

dayatsin-amd replied:
Both are valid. I changed it to iolink->node_to_id to be more efficient

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+{
+	struct tp_p2pgroup *group;
+
+	group = xmalloc(sizeof(*group));
```
You could use xzalloc and remove the memset below.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+
+struct tp_iolink *node_add_iolink(struct tp_node *node, uint32_t type, uint32_t node_to_id)
+{
+	struct tp_iolink *iolink = xmalloc(sizeof(*iolink));
```
You could use xzalloc and remove the memset below.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+{
+	struct tp_iolink *iolink = xmalloc(sizeof(*iolink));
+	if (!iolink) {
+		pr_err("Failed to allocate memory for iolink\n");
```
xalloc already prints an error. You probably don't need another error message here.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+{
+	struct tp_node *node = NULL;
+
+	node = xmalloc(sizeof(*node));
```
Use xzalloc, remove the memset and the error message.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+	memset(field, 0, sizeof(field));
+
+	p_begin = line;
+	while (p_begin && *p_begin == ' ')
```
Can p_begin be NULL and not cause a crash later on? That check seems pointless.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+		return false;
+
+	strncpy(field, p_begin, p_end - p_begin);
+	if (sscanf(field, "%lu", value) != 1)
```
I think you could save yourself the strncpy and just use sscanf(p_begin, "%lu", value). Or even simpler, replace all of the manual parsing above with one sscanf command:

    if (sscanf(p_begin, " %29s %lu", name, value) != 2) ...

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+
+		if (!dev->gpu_id && dev->cpu_cores_count >= 1) {
+			/* This is a CPU - we do not need to parse the other information */
+			continue;
```
"continue" at the end of the loop is pointless. Did you mean "break"?

dayatsin-amd replied:
Yes :) ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+				if (!strcmp(name, "size_in_bytes")) mem_size = value;
+			}
+
+			fclose(file);
```
If you want the "if (file)" check at the "fail:" label to be reliable, you need to set file = NULL here.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+				if (!strcmp(name, "type")) iolink_type = (uint32_t) value;
+				if (!strcmp(name, "node_to")) node_to_id = (uint32_t) value;
+			}
+			fclose(file);
```
file = NULL;

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+
+const char* p2pgroup_to_str(struct tp_p2pgroup *group)
+{
+	static char topology_printstr[60];
```
60 will be too small with more than 7 GPUs.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+
+const char* mapping_list_to_str(struct list_head *node_list)
+{
+	static char topology_printstr[60];
```
60 will get too small in the future if we ever have more than 8 GPUs.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+			if (iolink->type == TOPO_IOLINK_TYPE_XGMI) {
+				group = sys_get_p2pgroup_with_gpu_id(sys, iolink->type, node->gpu_id);
+				if (!group) {
+					group = sys_add_group(sys, iolink->type);
```
Before you create a new group, shouldn't you check whether there is already a group with peer_node in it? If there is, you'd want to add node to the existing group.

dayatsin-amd replied:
I don't think that is necessary. If this loop:
list_for_each_entry(node, &sys->nodes, listm_system) {

iterated over the peer_node before this node, then both this node and peer_node would already be in a group so we will not enter these two if statements:
if (!group) {

if (!p2pgroup_get_node_by_gpu_id(group, peer_node->gpu_id))

(this assumption only works for XGMI links because they are always bi-directional)

fxkamd replied:
Right, it works because the first node in the group has links to all the other members.

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+				continue;
+			}
+
+			len = read(fd, read_buf, sizeof(read_buf));
```
I don't think "read" gives you a 0-terminated string. So you should probably use sizeof(read_buf)-1 and set read_buf[len] = 0 on success.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+{
+	if (!src->valid)
+		return true;
+
```
What about dest->valid? If dest is invalid, I guess it shouldn't match anything in the source, so that should return false.

dayatsin-amd replied:
ACK

fxkamd replied:
I don't see any change for this ACK.
Edit: maybe I'm getting confused. I see our conversation here, but it's applied to your updated commit, not the original one. I think you added the "if (!dest->valid)" condition below. But the return value is the opposite of what I expected.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+				return false;
+			}
+		} else {
+			/* If GPUs have P2P-PCIe iolinks to this GPU, then at least one CPU will
```
You only parse io_links. Those do not include PCIe P2P links. They are only exposed under p2p_links on the DKMS branch. For now you can ignore these. But we are working on making that upstreamable so this will eventually land on amd-staging-drm-next. The existence of P2P PCIe links will depend on many factors. Just the existence of CPU->GPU links is not sufficient.

However, you probably should check XGMI P2P links between GPUs. But I guess you do that indirectly by matching XGMI hives below.

dayatsin-amd replied:
> You only parse io_links. Those do not include PCIe P2P links. They are only exposed under p2p_links on the DKMS branch. For now you can ignore these. But we are working on making that upstreamable so this will eventually land on amd-staging-drm-next. The existence of P2P PCIe links will depend on many factors. Just the existence of CPU->GPU links is not sufficient.

Yes, I put that comment because I felt that we were missing some cases here, but I did not have an easy way to test any code I would implement here. I added some words to the comment to make it more clear that there is probably more work to do here later.

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+	return "Unsupported";
+}
+
+struct tp_node *p2pgroup_get_node_by_gpu_id(const struct tp_p2pgroup *group,
```
Most functions here should be static because they are only used in this source file.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+		if (map_devices(src_sys, dest_sys, &src_nodes, &dest_nodes, maps)) {
+			list_del(&dest_group->listm_system);
+			pr_debug("Matched destination group [%s]\n", p2pgroup_to_str(dest_group));
+			if (match_xgmi_groups(src_sys, dest_sys, src_xgmi_groups, dest_xgmi_groups, maps)) {
```
Is this recursion intentional? Maybe a better explanation of how this is intended to work needed.

dayatsin-amd replied:
Yes. Added comments to explain recursion

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -0,0 +1,1104 @@
+			if (maps_append(maps, &new_maps))
+				return -EINVAL;
+
+			if (map_devices(src_sys, dest_sys, src_nodes, dest_nodes, maps)) {
```
Is this recursion intentional? Maybe a better explanation of how this is intended to work needed.

dayatsin-amd replied:
Yes. Added comments to explain recursion

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -703,3 +832,347 @@ int topology_parse(struct tp_system *sys, const char *message)
+	if (!src->valid)
+		return true;
+
+	if (!dest->valid)
```
If !dest->valid, I think that should not be considered a match.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -703,3 +832,347 @@ int topology_parse(struct tp_system *sys, const char *message)
+				pr_debug("Nodes after [0x%04X -> 0x%04X] did not match, adding list back\n",
+								src_node->gpu_id, dest_node->gpu_id);
+
+				list_add(&dest_node->listm_mapping, dest_nodes);
```
This adds dest_node back at the end of the list. That means you're going to see it again later in this loop. I think you need to add it back in the original position. Otherwise it can break the recursion. E.g. imagine dest_nodes is A-B-C-D. Then you recurse:

Level 1: take node A off the list, list_for_each_entry_save stores pointer to next entry: dest_node_temp = B
  Level 2: take node B off the list, next is C
    ...
    add B back at the end of the list. dest_nodes is now D-C-B
  back in level 1, add A back to the end of the list, dest_nodes is now: D-C-B-A

next loop iteration uses B, which was stored in dest_node_tmp, which is now near the end of the list. D and C are skipped and never considered.

fxkamd replied:
Never mind. list_adds to the front, not the end. So I think it works OK.

dayatsin-amd replied:
Yes. There is one of unit tests in a further commit that tests this situation.

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ * a GPU with large BAR on restore.
+ *
+ * GPUs can be part of 2 groups at the same time, i.e have P2P-XGMI links with one set of GPUs and
+ * have P2P-PCIe links with another set of GPUs.
```
I would not collect GPUs into groups using PCIe P2P. It gets a bit confusing. For example, when you have two XGMI hives with 4 GPUs each, and PCIe links between the two hives. How would you structure the groups?

XGMI groups: (G1,G2,G3,G4), (G5,G6,G7,G8)

G1 is connected via PCIe to G5,G6,G7,G8. So do you make a new group (G1,G5,G6,G7,G8) or separate groups (G1,G5), (G1,G6), (G1,G7), (G1,G8)? The first one means, you have GPUs in the group that are not connected to each other via PCIe. The second one means, each GPU is part of 5 groups (in this example).

dayatsin-amd replied:
This is an old comment and I forgot to remove it.

Initially, I was planning to implement it the way you mentioned (G1, G5, G6, G7, G8) but when I saw that each GPU would actually be part of 5 groups instead of 2, I thought it would make things too complicated. 
For now, it seems that the current code (only considering XGMI groups) can handle all uses cases ok, so I left it as is. I think we may have to re-visit this when we start looking at p2p_links with the dkms branch.

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ *      EXPECT: SUCCESS
+ *
+ * Test 2: 8 GPUs in 2 XGMI hives (partial P2P-PCIe case 1)
+ *	Same as test 2 but each hive have different device_id's so it is not possible to swap
```
"Same as test 2"? This is test 2.

dayatsin-amd replied:
:) ACK

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ *	6 GPUs have P2P-PCIe links and P2P-XGMI links
+ *	2 GPUs only have P2P-XGMI links
+ *	4 GPUs with P2P-PCIe link in first XGMI hive, 2 GPUs in the second XGMI hive
+ *      EXPECT: SUCCESS
```
You're just describing one topology. Does that mean you're remapping the topology onto itself?

I find all the descriptions below a bit strange, because they describe one topology, and then maybe some changes that are supposed to break the remapping. But it's not always clear what those changes are. It would be clearer if you defined a set of topologies, and then created tests that try to remap one topology to a different one, with an expected outcome of SUCCESS or FAIL.

This could also make the test code itself more compact and easier to understand. Write a helper function to create each of the topologies. Then each test would be very simple:

    create_topo_A(&tp_src);
    create_topo_B(&tp_dest);
    ret = set_restore_gpu_maps(&tp_src, &tp_dest, &maps);
    /* check results, verify maps etc. ... */

dayatsin-amd replied:
ACK I changed the test descriptions to be more clear about what when the source/destination are same or different.

I did not create general helper functions to create the topologies because there are only 1 or 2 cases where we re-use the same topology. Most of the time there are some nuances between each topology so we would not be able to re-use the helper functions.

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ *	Only 2 GPU's have P2P PCIe links
+ * 	Both GPUs with P2P-PCIe links are in different XGMI hives but at different index between
+ *      checkpointed and restored nodes
+ *      Only possible way to map is to map A002 -> B0000 and A017 -> B015
```
It's not clear what those numbers mean. How do you get to 17 with only 8 GPUs?

dayatsin-amd replied:
ACK, added some descriptions in the comments at beginning of file.

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ *      EXPECT: SUCCESS
+ *
+ * Test 4:8 GPU's in 2 XGMI hives (partial P2P-PCIe case 2)
+ *	Same as 3 but not possible to map because of CPU iolink changed
```
What changes about CPU iolinks? I'm assuming this is about links from CPU to GPU, which are missing if the GPU is small-BAR?

dayatsin-amd replied:
ACK - re-wrote the description

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ *
+ * Test 5: 8 GPUs with 1 XGMI hive
+ *    	4 GPUs in 1 XGMI hive, 4 GPUs have no XGMI bridge
+ *    	All 8 GPUs have P2P-PCIe links
```
This statement is not clear. PCIe P2P links to how many other GPUs? The GPUs in the XGMI hive won't have PCIe P2P links to other GPUs in the same hive.

dayatsin-amd replied:
Re-wrote description

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+ *
+ * Test 6: 5 GPUs (mix and match GPU types and partial large BAR)
+ *	No XGMI bridges
+ *	First 4 GPUs have P2P-PCIe links
```
P2P PCIe links to which other GPUs? They can only have links to large-BAR GPUs. But there are other factors that can prevent PCIe P2P links.

dayatsin-amd replied:
ACK - stopped using large BAR but P2P-PCIe links instead

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+                for (int j = 1; j < 4; j++)
+                        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_XGMI, ((i + j) % 4) + 2);
+                for (int j = 4; j < 8; j++)
+                        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_PCIE, ((i + j) % 4) + 6);
```
This adds links to nodes that haven't been created yet. Does that work?

Also, the %4 + 6 is confusing here. I think this would be clearer like this:

    for (int j = 6; j < 10; j++)
        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_PCIE, j);

dayatsin-amd replied:
>This adds links to nodes that haven't been created yet. Does that work?
Yes, Node is created 8 lines above: 
Line 139: node_gpus[i] = sys_add_node(&tp_src, i + 2, 0xA000 + i);

>Also, the %4 + 6 is confusing here. I think this would be clearer like this:
ACK

fxkamd replied:
> Yes, Node is created 8 lines above:

No. You're creating links to node 6-9 (the second hive) here. Those are created below in the "for (int i = 4; i < 8; i++)" loop.

dayatsin-amd replied:
>Yes, Node is created 8 lines above:
>No. You're creating links to node 6-9 (the second hive) here. Those are created below in the "for (int i = 4; i < 8; i++)" loop.

That is ok. We can create links to node that do not exist yet because the links are validated later during topology_determine_iolinks(). It is only the source node that has to exist at the time the link is created. iolinks are stored with the source node.

fxkamd commented on plugins/amdgpu/tests/test_topology_remap.c:
```diff
@@ -0,0 +1,985 @@
+                for (int j = 1; j < 4; j++)
+                        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_XGMI, ((i + j) % 4) + 6);
+                for (int j = 4; j < 8; j++)
+                        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_PCIE, ((i + j) % 4) + 2);
```
Similar to above, make this

    for (int j = 2; j < 6; j++)
        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_PCIE, j);

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -13,6 +13,7 @@
 #include <sys/mman.h>
 #include <sys/types.h>
 #include <stdint.h>
+#include <pthread.h>
```
Don't you need to add -lpthread to the compiler command line in the Makefile somewhere?

dayatsin-amd replied:
ACK. Forgot to git add Makefile and the change went into another commit in the other PR

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -528,6 +529,116 @@ void amdgpu_plugin_fini(int stage, int ret)
+			} /* PROCPIDMEM read done */
+               }
+        }
+	pr_info("amdgpu_plugin: Thread[0x%x] done num_bos:%d ret:%d\n", thread_data->gpu_id, num_bos, ret);
```
Indentation is off. It looks like your editor is set to tab-width 4 and getting mixed up when to indent with tabs or spaces. For kernel code (and I think CRIU uses the same convention), TAB width should be 8 and we should always use TABs for indentation where possible.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -528,6 +529,116 @@ void amdgpu_plugin_fini(int stage, int ret)
+        return NULL;
+failed:
+        pr_info("amdgpu_plugin: Thread[0x%x] failed ret:%d\n", thread_data->gpu_id, ret);
+	thread_data->ret = ret;
```
Bad indentation.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -528,6 +529,116 @@ void amdgpu_plugin_fini(int stage, int ret)
+                       continue;
+
+		num_bos++;
+		if (bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_VRAM ||
```
To reduce indentation, I'd suggest

    if (!(bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_VRAM) &&
        !(bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_GTT))
        continue;

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -528,6 +529,116 @@ void amdgpu_plugin_fini(int stage, int ret)
+			int mem_fd;
+
+			if (bo_info_test[i]->bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_PUBLIC) {
+				int drm_fd;
```
To reduce indentation further, you could move the two different dumping implementations (map BO or /proc/pid/mem) each into its own helper functions. But that's up to you.

dayatsin-amd replied:
I like this idea as I think the functions in this file are generally too big, which in turn causes a lot of identation.

I will change it after as I think there was some additional optimizations that can be done to these loops:
1. For mmap section. we do not need to call open_drm_render_device for each BO. (can just mmap with different offsets)
2. For /proc/pid/mem section, do not need to fopen for each BO and can just call lseek with different offsets.

I am making some further changes to this function as part of the fix for tensorflow (protobuf cannot handle buffers > 4GB).

And I also started implemented multi-threading for the restore section so I will have more changes to this commit.

dayatsin-amd committed:
criu/plugin: Restore libhsakmt shared memory files

Libhsakmt(thunk) uses a shared memory file in /dev/shm/hsakmt_shared_mem
and its semaphore in /dev/shm/hsakmt_shared_mem. Adding a check during
checkpoint to see if these two files exist. If they exist then the
plugin will try to restore them during restore.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -432,15 +442,76 @@ int devinfo_to_topology(DevinfoEntry *devinfos[], uint32_t num_devices, struct t
+		pr_info("param: KFD_VRAM_SIZE_CHECK:%s\n", kfd_vram_size_check ? "Y" : "N");
+
+		if ((opt_param = getenv("KFD_IGNORE_NUMA"))) {
+			if (!strcmp(opt_param, "1") || !strcmp(opt_param, "Y"))
```
This is a bit inconsistent. The other parameters accepted "0" or "NO". This one accepts "1" or "Y". Should that be "YES" to be consistent with "NO"?

That said, why does this option have the opposite logic? Why not call it "KFD_NUMA_CHECK" with default "1"? That would make it behave the same as all the other options.

I would also move the parsing of boolean parameters into a helper function to avoid all that duplication and minimize the risk of introducing more inconsistencies later.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin_topology.c:
```diff
@@ -1098,6 +1135,63 @@ bool match_xgmi_groups(struct tp_system *src_sys, struct tp_system *dest_sys,
+/* return 1  	 if gpu override is set
+ * return 0  	 if no gpu overide
+ * return -errno if failed to set gpu override */
+int get_user_gpu_override(struct tp_system *src_sys, struct tp_system *dest_sys,
```
If I understand this correctly, it applies very different rules for the override. For example it doesn't do any XGMI hive matching. And it doesn't have any flexibility in remapping source GPUs to a compatible destination GPU. It has to match in the order specified.

Maybe it would be easier, more flexible and more correct to use KFD_DESTINATION_GPUS as a filter that specifies which GPUs are available for mapping. I.e. just remove the GPUs not specified from dest_sys. Then you can use the normal match_xgmi_groups function to do the actual matching, but with a subset of the GPUs.

dayatsin-amd replied:
The intention of this parameter was to allow the user to force the GPUs to be mapped in a certain order. So for example if the GPUs were A1, A2, A3, A4 on the source node and the mapping code would map them to B1, B2, B3, B4, but for unforeseen reasons, the user would want the GPUs to be mapped to B3, B4, B2, B1.
With current code, the number of GPUs on destination node needs to match number of GPUs on source node. If the destination node has more GPUs, the user is expected to mask the extra GPUs using cgroups.

The check here:
```
/* Ignore extra GPU's */
if (index >= dest_sys->num_nodes)
    break;
    
```
Is in case a user specified: KFD_DESTINATION_GPUS=0x0001, 0x0002, 0x0003, 0x0004 (for e.g because they have a generic script for all their systems), but there are only 3 GPUs in the system.

fxkamd replied:
That's weird. The only way that you can have index >= dest_sys->num_nodes is, if some of the GPU names are invalid. So shouldn't that be an error?

Another prossible error is, if the same GPU is specified multiple times. Is there a way to catch such an error?

Masking GPUs with cgroups won't work for the customer running in a privileged docker container. So that means they'll be forced to use KFD_DESTINATION_GPUS. We need to make this as robust as possible (i.e. catch as many user errors as possible). That includes:

- bad GPU IDs
- same GPU specified multiple times
- specified list of destination GPUs incomatible with the source link topology and XGMI topology

Duplicating all the link compatibility checking in the override function is error prone and harder to maintain. Therefore I would strongly recommend, relying on as much common code as possible. If that means letting the common code reorder the destination GPUs, I think that's OK.

dayatsin-amd replied:
To be honest, being able to re-order the GPU's was the main reason why I implemented this variable (I used it mostly for internal testing/debugging) and it was meant to be more of a power-user feature, which is why I did not implemented too many validation checks. But I can see the use case for the way you want to use it and also how the environment variable name could be misleading. Would you be ok if I implement this:

KFD_DESTINATION_GPUS: List of allowed GPUs on restore node. GPUs in this list may be re-ordered.
KFD_OVERRIDE_GPUS: Will be used to force a specific GPU order.

For both cases, I will add checks for invalid and duplicate GPU ID's and/or minor render numbers.

Note: Current code will catch case where destination GPU does not have same XGMI topology as source GPU. In map_device function we go through list of iolinks and call iolink_match() function to compare them.

dayatsin-amd replied:
I removed this feature :)

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -848,6 +856,48 @@ int amdgpu_plugin_dump_file(int fd, int id)
+
+		dev = sys_get_node_by_gpu_id(&src_topology, thread_datas[i].gpu_id);
+		if (!dev) {
+			ret = -EFAULT;
```
EFAULT means "bad address". Is that really the appropriate error code? Maybe ENODEV.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -848,6 +856,48 @@ int amdgpu_plugin_dump_file(int fd, int id)
+
+		thread_datas[i].drm_fd = open_drm_render_device(dev->drm_render_minor);
+		if (thread_datas[i].drm_fd < 0) {
+			ret = -EFAULT;
```
Doesn't drm_fd already contain an error code? Why not assign that to ret?

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -848,6 +856,48 @@ int amdgpu_plugin_dump_file(int fd, int id)
+		ret_thread = pthread_create(&thread_datas[i].thread, NULL, dump_bo_contents, (void*) &thread_datas[i]);
+		if (ret_thread) {
+			pr_err("Failed to create thread[%i]\n", i);
+			ret = -EINVAL;
```
ret_thread already contains a perfectly good error code. You just need to negate it.

    ret = -ret_thread;

Or you could just do this above:

    ret = -pthread_create(...);

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -848,6 +856,48 @@ int amdgpu_plugin_dump_file(int fd, int id)
+			close(thread_datas[i].drm_fd);
+
+		if (thread_datas[i].ret) {
+			ret = -EIO;
```
You could assign an appropriate error code to thread_datas[i].ret in dump_bo_contents. Then you don't have to make one up here.

dayatsin-amd replied:
ACK

fxkamd reviewed (commented):
Some nit-picks about error codes. I see some inconsistent or inappropriate error codes in other plugin code as well, but missed it in previous reviews.

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -1454,115 +1559,40 @@ int amdgpu_plugin_restore_file(int id)
+		ret_thread = pthread_create(&thread_datas[i].thread, NULL, restore_bo_contents, (void*) &thread_datas[i]);
+		if (ret_thread) {
+			pr_err("Failed to create thread[%i]\n", i);
+			fd = -EBADFD;
```
Use -ret_thread for consistency.

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -1454,115 +1559,40 @@ int amdgpu_plugin_restore_file(int id)
 			close(devinfo_bucket_ptr[i].drm_fd);
+
+		if (thread_datas[i].ret) {
+			fd = -EBADFD;
```
Assign an appropriate error code in the thread function. Then you don't have to make one up here.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -729,6 +735,87 @@ void *restore_bo_contents(void *_thread_data)
+	*shared_mem_size = st.st_size;
+
+	/* First 4 bytes of shared file is the magic */
+	memcpy(shared_mem_magic, shared_mem_ptr, sizeof(*shared_mem_magic));
```
You can do this with a normal assignment:

    *shared_mem_magic = *shared_mem_ptr;

But that said, is it really worth doing this whole mmap dance for copying a single value? Why not just use a read system call?

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -729,6 +735,87 @@ void *restore_bo_contents(void *_thread_data)
+			return -EINVAL;
+		}
+
+		shared_mem_ptr = mmap(0, shared_mem_size, PROT_WRITE, MAP_SHARED, fd, 0);
```
The first parameter of mmap is a pointer. By convention it should be NULL, not 0. Also, creating a write-only mapping is a bit unusual. There is probably no harm in making it PROT_READ | PROT_WRITE.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -729,6 +735,87 @@ void *restore_bo_contents(void *_thread_data)
+			close(fd);
+			return -EACCES;
+		}
+		memcpy(shared_mem_ptr, &shared_mem_magic, sizeof(shared_mem_magic));
```
Use an assignment here:

    *shared_mem_ptr = shared_mem_magic;

Or avoid the whole mmap trouble and just use a write system call to write the magic number to the start of the file.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -729,6 +735,87 @@ void *restore_bo_contents(void *_thread_data)
+		return 0;
+	}
+
+	shared_mem_ptr = mmap(0, st.st_size, PROT_READ, MAP_SHARED, fd, 0);
```
The first parameter of mmap is a pointer. By convention it should be NULL, not 0.

dayatsin-amd replied:
ACK

fxkamd commented on Documentation/amdgpu_plugin.txt:
```diff
@@ -35,6 +35,82 @@ Dependencies
+
+    KFD_DESTINATION_GPUS=65329,37083
+
+    KFD_DESTINATION_GPUS=renderD129,renderD128
```
You removed this feature. Remove the comment as well.

dayatsin-amd replied:
ACK

fxkamd commented on Documentation/amdgpu_plugin.txt:
```diff
@@ -35,6 +35,82 @@ Dependencies
+    Enable or disable topology check.
+    If enabled, amggpu_plugin will check whether the destination GPU has
+    sufficient capabilities to run checkpointed application. This option is
+    only valid if KFD_DESTINATION_GPUS is used. Default:Enabled
```
Remove the comment about KFD_DESTINATION_GPUS. If this depends on KFD_DESTINATION _GPUS, should you remove KFD_TOPOLOGY_CHECK as well?

dayatsin-amd replied:
ACK - correct. KFD_TOPOLOGY_CHECK is removed as well
